packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -394,6 +394,52 @@ export default function parse(source: string, state: State): TemplateParseResult
         element.lwc = lwcOpts;
     }
 
+    function applyLwcRenderModeDirective(element: IRElement) {
+        const { parent, tag, attrsList } = element;
+
+        // Can't handle slots in applySlot because it would be too late for class and style attrs
+        if (state.renderMode === LWCDirectiveRenderMode.light && tag === 'slot') {
+            const invalidAttrs = attrsList
+                .filter(({ name }) => name !== 'name')
+                .map(({ name }) => name);
+            if (invalidAttrs.length > 0) {
+                return warnOnElement(
+                    ParserDiagnostics.LWC_LIGHT_SLOT_INVALID_ATTRIBUTES,
+                    element.__original,
+                    [invalidAttrs.join(',')]
+                );
+            }
+        }
+
+        if (parent !== undefined) return;
+
+        const lwcRenderModeAttribute = getTemplateAttribute(
+            element,
+            ROOT_TEMPLATE_DIRECTIVES.RENDER_MODE
+        );
+
+        if (!lwcRenderModeAttribute) {
+            return;
+        }
+
+        removeAttribute(element, ROOT_TEMPLATE_DIRECTIVES.RENDER_MODE);
+
+        if (
+            lwcRenderModeAttribute.type === IRAttributeType.String &&
+            !hasOwnProperty.call(LWCDirectiveRenderMode, lwcRenderModeAttribute.value)
+        ) {
+            const possibleValues = Object.keys(LWCDirectiveRenderMode)
+                .map((value) => `"${value}"`)
+                .join(', or ');
+            warnOnElement(ParserDiagnostics.LWC_RENDER_MODE_INVALID_VALUE, element.__original, [
+                possibleValues,
+            ]);
+            return;
+        }
+
+        state.renderMode = lwcRenderModeAttribute.value as LWCDirectiveRenderMode;
```

Comment:
It feel wrong to change the `renderMode` in the middle of the AST traversal. Since this information is present on the root element, it would make more sense to collect it before traversing the AST tree.